text/0000-list-rendering-reform.md

```diff
@@ -0,0 +1,267 @@
+- There is no way to easily render based on even vs odd list items.
+
+### Prior art
+
+- [Vue `v-for`](https://vuejs.org/guide/essentials/list.html)
+- [Svelte `{#each}`](https://svelte.dev/docs/logic-blocks#each)
+- [Angular `*ngFor`](https://angular.io/api/common/NgForOf#local-variables)
+- [Solid `<For>`](https://docs.solidjs.com/references/api-reference/control-flow/For)
+- [Lit `repeat`](https://lit.dev/docs/templates/lists/#the-repeat-directive)
+
+## Detailed design
+
+The design of `lwc:each` is almost identical to that of `for:each`. There are only a few modifications.
+
+First, all directives are prefixed with `lwc:`:
+
+- `for:each` → `lwc:each`
+- `for:item` → `lwc:item`
+- `key` → `lwc:key`
+- `for:index` → `lwc:index`
+
+Second, the `lwc:key` must be declared on the iterator root, not on each item within the loop:
+
+```html
+<template lwc:each={items} lwc:item="item" lwc:key={item.id}>
+    {item.name}
+</template>
+```
+
+Third, four new convenience directives are added:
+
+- `lwc:first` - boolean that is `true` if the item is first in the list
+- `lwc:last` - boolean that is `true` if the item is last in the list
+- `lwc:even` - boolean that is `true` if the item index is even
+- `lwc:odd` - boolean that is `true` if the item index is odd
+
+Each directive will be detailed separately.
+
+### `lwc:each`
+
+This functions nearly identically to `for:each`. (That's kind of the point – it should not be difficult for component authors to migrate.) Like `for:each`, it is supported in both the `<template>` format:
+
+```html
+<template lwc:each={items} lwc:item="item" lwc:key={item.id}>
+    {item.name}
+</template>
+```
+
+…as well as the shorthand, `<template>`-less format:
+
+```html
+<li lwc:each={items} lwc:item="item" lwc:key={item.id}>
+    {item.name}
+</li>
+```
+
+Identically to `for:each`, it supports any [Iterable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols).
+
+### `lwc:item`
+
+Identical to `for:item`. It is required for `lwc:each`, and throws a compile-time error if it is not defined:
+
+    lwc:each and lwc:item directives should be associated together.
+
+(This is the same error thrown for `for:each` in the same situation.)
+
+[Unlike `for:item`](https://stackblitz.com/edit/salesforce-lwc-m56edb?file=src%2Fmodules%2Fx%2Fapp%2Fapp.html,src%2Fmodules%2Fx%2Fapp%2Fapp.js&title=LWC%20playground), `lwc:item` cannot use the same variable name as `lwc:each`:
+
+```html
+<template lwc:each={foo} lwc:item="foo" lwc:key={foo.id}>
+</template>
+```
+
+The above will throw a compile-time error, since otherwise the behavior may be unexpected and confusing, especially when considering the `lwc:key` on the same element.  (Which `foo` does `lwc:key={foo.id}` refer to above? Rather than guessing, we will throw a compile-time error.)
+
+### `lwc:key`

```

Comment:
One usecase that is not very clear about existing `for:each` is nesting iteration blocks, where the inner `for:each` directive is an immediate child of the outer one. What should be the behavior of the `lwc:key` in that case?
IMO, this semantic makes it more clear than the older one. One thread related this confusion 
[1](https://salesforce.stackexchange.com/questions/372324/how-can-i-do-a-nested-template-foreach-in-lwc-for-a-map-of-arrays)

Reply:
Yes, exactly. With `lwc:key` on the same element as `lwc:each`, you know exactly where the key should be applied.

Reply:
I think we might need to slightly distinguish the way `lwc:key` is applied vs `key`.  With `for:each` when the short-hand is used, the `key` is applied to the element containing the `for:each`:

```html
<li lwc:each={items} lwc:item="item" lwc:key={item.id}>
    {item.name}
</li>
```

In the example, the key is applied to the `li` and not the `{item.name}`, whereas if it were a `template` the `key` would be applied to the children inside the `template`.

The reason I mention this is because when there is a nested loop where the outer loop is a `template` the `key` is only applied once:

[stackblitz example (playground link)](https://stackblitz.com/edit/salesforce-lwc-2rsr94?file=src%2Fmodules%2Fx%2Fapp%2Fapp.html)

```html
<template>
    <template for:each={contacts} for:item="contact">
        <div for:each={contact.addresses} for:item="address" key={contact.id}>
            <section>{address.zip}</section>
        </div>
</template>
```
Note in the `for:each` example, the `key` only appears on the `div`, applying the `key` on the `section` would produce a warning.

If I understood correctly, with `lwc:each` the new format would be:

```html
<template>
    <template lwc:each={contacts} lwc:item="contact" lwc:key={contact.id}>
        <div lwc:each={contact.addresses} lwc:item="address" lwc:key={address.id}>
            <section>{address.zip}</section>
        </div>
</template>
```

How would the `lwc:key={address.id}` be applied in this scenario, on the `div` or on the `section`?

Reply:
@jmsjtu I think the example you provide is a bug in `for:each`. It doesn't make any sense to me that one loop _must_ have a `key`, but the other loop _must not_ have a `key`.

In the case you mention, I think the problem arises from the ambiguity of what, exactly, the `key` applies to:

```html
<template for:each={contacts} for:item="contact">
    <div key={contact.id} for:each={contact.addresses} for:item="address"></div>
</template>
```

In the above example, does the `key` correspond to the outer (non-shorthand) `template` loop? Or to the inner (shorthand) `div` loop? It's ambiguous.

Another benefit of this RFC, IMO, is that `lwc:key` is never ambiguous. It always applies to exactly the loop on the same element, regardless of shorthand or no-shorthand.